### PR TITLE
User Authentication

### PR DESCRIPTION
**Pull request by anim-13**
Implemented User Authentication . i.e created user sign up and user log in/out page.

Trello Ticket Link:
<blockquote class="trello-card"><a href="https://trello.com/c/2wLPdpl3/10-e-state-website-user-authentication-flow">E-state Website || User Authentication Flow</a></blockquote><script src="https://p.trellocdn.com/embed.min.js"></script>

Screenshot from 2021-03-05 16-46-22
Screenshot from 2021-03-05 16-46-32

### CODE REVIEW

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (kalpana19sharma) [inline app/assets/stylesheets/sessions.scss]: Remove this unnecessary file

File: app/assets/stylesheets/users.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (kalpana19sharma) [inline app/assets/stylesheets/users.scss]: Remove this unnecessary file

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,10 @@
 class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+  helper_method :current_user
+  
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
+  end
+
 end
+
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: Remove unnecessary empty lines.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,10 @@
 class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+  helper_method :current_user
+  
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb:6]: 1. What does `||=` operator do?
2. What's the difference between `find` and `find_by`

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb:3]: Remove unnecessary empty line.

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  
+  def new
+  end
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: For empty methods we use single line syntax 
```suggestion
  def new; end
```

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      session[:user_id] = user.id.to_s
+      redirect_to root_path, notice: 'Successfully logged in!'
+    else
+      flash.now.alert = "Incorrect email or password, try again."
+      render :new
+    end
+  end
+
+  def destroy
+
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Remove unnecessary empty line.

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      render :new
+    end
+  end
+
+  def destroy
+    
+    session.delete(:user_id)
+    redirect_to login_path, notice: "Logged out!"
+  end
+
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Remove unnecessary empty line.

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    end
+  end
+
+  def destroy
+    
+    session.delete(:user_id)
+    redirect_to login_path, notice: "Logged out!"
+  end
+  
+end
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Add an empty line at the end of the file.

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:login][:email].downcase)
+    
+    if user && user.authenticate(params[:login][:password])
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Use strong parameters

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Remove unnecessary empty line

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+  
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    @user.email.downcase!
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: What's this for?

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    if @user.save
+      flash[:notice] = "Account created successfully!"
+      redirect_to root_path
+    else
+      flash.now.alert = "Enter valid email and password and try again."
+      render :new
+    end
+  end
+
+private
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Fix indentation

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+  end
+  
+end
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Add an empty line at the end of the file.

File: app/helpers/sessions_helper.rb [hunk 15]
```diff
@@ -0,0 +1,2 @@
+module SessionsHelper
+end
```
Comment (kalpana19sharma) [inline app/helpers/sessions_helper.rb]: Remove unnecessary file

File: app/helpers/users_helper.rb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
+end
```
Comment (kalpana19sharma) [inline app/helpers/users_helper.rb]: Remove unnecessary file.

File: app/models/user.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+	has_secure_password
+	validates :email, presence: true, uniqueness: true
+	
+end
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Fix indentation

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -11,6 +11,21 @@
...
+    <% end %>
+    <hr>
+	
+    <% if notice %>
+      <%= notice %>
+    <% end %>
+	
+    <% if alert %>
+      <%= alert %>
+    <% end %>
```
Comment (kalpana19sharma) [inline app/views/layouts/application.html.erb]: You can extract this logic in a partial

File: app/views/sessions/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+    <%= f.text_field :email, autofocus: true %>
+  </div>
+  <div>
+    <%= f.label :password %><br>
+    <%= f.password_field :password %>
+  </div>
+  <div>
+    <%= f.submit "Log In" %>
+  </div>
+<% end %>
```
Comment (kalpana19sharma) [inline app/views/sessions/new.html.erb]: empty line.

File: app/models/user.rb [hunk 20]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+	has_secure_password
+	validates :email, presence: true, uniqueness: true
```
Comment (kalpana19sharma) [inline app/models/user.rb]: No validations on password?

File: app/views/users/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+    <%= f.password_field :password %>
+  </div>
+  <div>
+    <%= f.label :password_confirmation %><br>
+    <%= f.password_field :password_confirmation %>
+  </div>
+  <div>
+    <%= f.submit "Sign up!" %>
+  </div>
+<% end %>
```
Comment (kalpana19sharma) [inline app/views/users/new.html.erb]: empty line.

File: config/routes.rb [hunk 22]
```diff
@@ -1,3 +1,11 @@
...
+
+	root to: 'users#new'
+
+	get 'users/new' => 'users#new', as: :new_user
+	post 'users' => 'users#create'
+	get '/login'     => 'sessions#new'
+	post '/login'    => 'sessions#create'
+	delete '/logout' => 'sessions#destroy'  
+  	
+end
```
Comment (kalpana19sharma) [inline config/routes.rb]: Fix indentation.

File: config/routes.rb [hunk 23]
```diff
@@ -1,3 +1,11 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
-end
+
+	root to: 'users#new'
+
+	get 'users/new' => 'users#new', as: :new_user
+	post 'users' => 'users#create'
```
Comment (kalpana19sharma) [inline config/routes.rb]: Can we use resources here?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]
@anim-13 please remove unused test files.

**Comment by kalpana19sharma** [comment]
@anim-13 add proper screenshots in the PR description

File: app/controllers/application_controller.rb [hunk 24]
```diff
@@ -1,2 +1,9 @@
...
+  end
 end
+
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: I think some whitespaces are added here.

File: app/controllers/sessions_controller.rb [hunk 25]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: session_params['email'])  
+    if user && user.authenticate(session_params['password'])
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: 1. Use symbol syntax
```suggestion
    if user && user.authenticate(session_params[:password]) 
```
2. Use `&.` operator

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -0,0 +1,22 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    @user.email.downcase!
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: What's this logic for? Isn't this should be in model?

File: app/models/user.rb [hunk 27]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+	has_secure_password
+	validates :email, presence: true, uniqueness: true	
+  validates :password, confirmation: true, length: {minimum: 6, maximum: 20}
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Fix indentation

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
+    @user.email.downcase!  
+    if @user.save
+      flash[:notice] = "Account created successfully!"
+      redirect_to root_path
+    else
+      flash.now.alert = "Enter valid email and password and try again."
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb:12]: If you are showing error messages on the form then I think you can skip the flash message

File: app/helpers/application_helper.rb [hunk 29]
```diff
@@ -1,2 +0,0 @@
-module ApplicationHelper
-end
```
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb]: Don't remove the files from initial commit.

File: app/views/layouts/_navigation.html.erb [hunk 30]
```diff
@@ -0,0 +1,7 @@
+<% if notice %>
+  <%= notice %>
+<% end %>	
+<% if alert %>
+  <%= alert %>
+<% end %>
+
```
Comment (kalpana19sharma) [inline app/views/layouts/_navigation.html.erb]: Whitesapces

File: app/views/layouts/_navigation.html.erb [hunk 31]
```diff
@@ -0,0 +1,7 @@
+<% if notice %>
+  <%= notice %>
+<% end %>	
+<% if alert %>
+  <%= alert %>
+<% end %>
```
Comment (kalpana19sharma) [inline app/views/layouts/_navigation.html.erb]: Can we refactor this logic?

File: app/views/layouts/application.html.erb [hunk 32]
```diff
@@ -10,7 +10,14 @@
...
 
-  <body>
+  <body>  
+    <% if current_user %>
+      <%= "Logged in as #{current_user.email}" %> | <%= link_to 'Home', root_path %> | <%= link_to 'Log Out', logout_path, method: :delete %>
+    <% else %>
+      <%= link_to 'Home', root_path %> | <%= link_to 'Log In', login_path %> or <%= link_to 'Sign Up', new_user_path %>
+    <% end %>
+    <hr>
+    <%= render "layouts/navigation"%>
```
Comment (kalpana19sharma) [inline app/views/layouts/application.html.erb]: Use proper spaces

File: app/views/layouts/_navigation.html.erb [hunk 33]
```diff
@@ -0,0 +1,7 @@
+<% if notice %>
```
Comment (kalpana19sharma) [inline app/views/layouts/_navigation.html.erb]: Why this filename is `navigation`?

File: config/database.yml [hunk 34]
```diff
@@ -20,6 +20,9 @@ default: &default
   # For details on connection pooling, see Rails configuration guide
   # https://guides.rubyonrails.org/configuring.html#database-pooling
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+  username: animesh
+  password: password
```
Comment (kalpana19sharma) [inline config/database.yml]: Don't commit the passwords

File: config/routes.rb [hunk 35]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+	root to: 'users#new'
+  resources :users
+	get    '/login'    => 'sessions#new'
+	post   '/login'    => 'sessions#create'
+	delete '/logout'   => 'sessions#destroy'
```
Comment (kalpana19sharma) [inline config/routes.rb]: Fix indentation

File: config/routes.rb [hunk 36]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+	root to: 'users#new'
+  resources :users
```
Comment (kalpana19sharma) [inline config/routes.rb:3]: Do we need all of the default routes?

File: test/controllers/sessions_controller_test.rb [hunk 37]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (kalpana19sharma) [inline test/controllers/sessions_controller_test.rb]: Remove unnecessary test files

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/views/layouts/_navigation.html.erb (hunk 33 above)
Comment (anim-13) [reply app/views/layouts/_navigation.html.erb]: changed to mesage.html.erb

File: app/controllers/sessions_controller.rb [hunk 38]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: session_params[:email])  
+    if user && user.authenticate(session_params[:password])
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Use `&.` operator

File: app/controllers/sessions_controller.rb [hunk 39]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: session_params[:email])  
+    if user && user.authenticate(session_params[:password]) 
+      session[:user_id] = user.id.to_s
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]: Do we really need to use `to_s` here?

File: app/controllers/sessions_controller.rb [hunk 40]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: session_params[:email])
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb:7]: Use a before filter to initialize the user.

File: app/helpers/application_helper.rb (hunk 29 above)
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb]: Restore this file.

Comment (kalpana19sharma) [reply app/helpers/application_helper.rb]: It was added in initial commit and not generated when you created a controller.

File: app/models/user.rb [hunk 41]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  before_save { email.downcase! }
+	has_secure_password
+	validates :email, presence: true, uniqueness: true	
+  validates :password, confirmation: true, length: {minimum: 6, maximum: 20}
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Mam it is showing correct in my editor. I don't know why it is still showing in this particular file.

File: config/routes.rb (hunk 36 above)
Comment (kalpana19sharma) [reply config/routes.rb:3]: ??

File: test/fixtures/users.yml [hunk 42]
```diff
@@ -0,0 +1,11 @@
...
+
+one:
+  name: MyString
+  email: MyString
+  password_digest: MyString
+
+two:
+  name: MyString
+  email: MyString
+  password_digest: MyString
```
Comment (kalpana19sharma) [inline test/fixtures/users.yml]: Remove this file

File: test/models/user_test.rb [hunk 43]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class UserTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (kalpana19sharma) [inline test/models/user_test.rb]: Remove this file

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb [hunk 44]
```diff
@@ -0,0 +1,21 @@
...
+      render :new
+    end
+  end
+
+  private
+    def user_params
+      params.require(:user).permit(:name, :email, :password, :password_confirmation)
+    end
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Fix indentation
```suggestion
  private
  def user_params
    params.require(:user).permit(:name, :email, :password, :password_confirmation)
  end  
```

File: app/helpers/application_helper.rb  [hunk 45]
```diff
@@ -1,2 +1,2 @@
 module ApplicationHelper
-end
+end
```
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb ]: No new line at the end of the file.

File: app/models/user.rb [hunk 46]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  before_save { email.downcase! }
+  has_secure_password
```
Comment (kalpana19sharma) [inline app/models/user.rb:4]: Add an empty line above `has_secure_password`

File: config/routes.rb [hunk 47]
```diff
@@ -1,3 +1,7 @@
...
+	get '/login' => 'sessions#new'
+	post '/login' => 'sessions#create'
+	delete '/logout' => 'sessions#destroy'
```
Comment (kalpana19sharma) [inline config/routes.rb]: Mam yesterday we had corrected this indentation error by using gitk. Yet it is still showing in github.

File: test/fixtures/users.yml (hunk 42 above)
Comment (kalpana19sharma) [inline test/fixtures/users.yml]: Remove this file.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**